src/internal/components/options-list/utils/use-highlight-option.ts

```diff
@@ -1,64 +1,58 @@
-import { MutableRefObject, useCallback, useState } from 'react';
+import { useCallback, useState } from 'react';
+
+export type HighlightType = 'keyboard' | 'mouse';
 
 export function createHighlightedOptionHook<OptionType>({
   isHighlightable,
 }: {
   isHighlightable: (option: OptionType) => boolean;
 }) {
-  return function useHighlightedOption({
-    options,
-    isKeyboard,
-  }: {
-    options: ReadonlyArray<OptionType>;
-    isKeyboard: MutableRefObject<boolean>;
-  }) {
+  return function useHighlightedOption({ options }: { options: ReadonlyArray<OptionType> }) {
     const [highlightedIndex, setHighlightedIndexState] = useState(-1);
-    const [highlightedType, setHighlightedType] = useState<'mouse' | 'keyboard'>(
-      isKeyboard.current ? 'keyboard' : 'mouse'
-    );
-    const setHighlightedIndex = useCallback(
-      (index: number) => {
-        setHighlightedIndexState(index);
-        setHighlightedType(isKeyboard.current ? 'keyboard' : 'mouse');
-      },
-      [isKeyboard]
-    );
+    const [highlightType, setHighlightType] = useState<HighlightType>('keyboard');
+    const setHighlightedIndex = useCallback((index: number, highlightType: HighlightType) => {
+      setHighlightedIndexState(index);
+      setHighlightType(highlightType);
+    }, []);
     const highlightedOption =
       options[highlightedIndex] && isHighlightable(options[highlightedIndex]) ? options[highlightedIndex] : undefined;
 
-    const moveHighlightFrom = (direction: -1 | 1, startIndex = highlightedIndex) => {
+    const moveHighlightFrom = (direction: -1 | 1, startIndex = highlightedIndex, highlightType: HighlightType) => {
       let newIndex = startIndex;
       do {
         newIndex += direction;
       } while (options[newIndex] && !isHighlightable(options[newIndex]));
 
       if (options[newIndex]) {
-        setHighlightedIndex(newIndex);
+        setHighlightedIndex(newIndex, highlightType);
       }
     };
 
-    const moveHighlight = (direction: -1 | 1) => moveHighlightFrom(direction);
+    const moveHighlight = (direction: -1 | 1, highlightType: HighlightType) =>
+      moveHighlightFrom(direction, highlightedIndex, highlightType);
 
     const highlightOption = useCallback(
-      (option: OptionType) => {
+      (option: OptionType, highlightType: HighlightType) => {
         const index = options.indexOf(option);
-        setHighlightedIndex(index);
+        setHighlightedIndex(index, highlightType);
       },
       [options, setHighlightedIndex]
     );
 
     return {
-      setHighlightedIndex,
+      highlightType,
       highlightedIndex,
-      highlightedType,
       highlightedOption,
-      moveHighlight,
-      resetHighlight: () => setHighlightedIndex(-1),
-      goHome: () => moveHighlightFrom(1, -1),
-      goEnd: () => moveHighlightFrom(-1, options.length),
-      highlightOption,
+      // Mouse handlers
+      setHighlightedIndex: (index: number) => setHighlightedIndex(index, 'mouse'),
```

Comment:
Do you think, we somehow should imply input method in the handler name? For example `setMouseHighlight` or `moveKeyboardHighlight`?

Reply:
Yeah, I was thinking about this as well but couldn't come up with some good names.

How about:
```
interface HighlightedOptionApi<OptionType> {
  highlightType: HighlightType;
  highlightedIndex: number;
  highlightedOption: OptionType;
  mouse: {
    highlightOption(index: number): void;
  },
  keyboard: {
    resetHighlight(): void;
    highlightOption(option: OptionType): void;
    moveHighlight(direction: -1 | 1): void;
    moveHighlightToStart(): void;
    moveHighlightToEnd(): void;
  }
}
```

Reply:
There is a slight concern about use of destructuring

```js
const { keyboard: { resetHighlight } } = useHiglightedOption()

// many lines of code

() => resetHighlight();
```

If you want, you may proceed with your current idea (it is not a public API, so always easy to revisit). But I prefer having `Keyboard` or `Mouse` embedded directly into the name